libfwupdplugin/fu-usb-device.c

```diff
@@ -267,13 +269,381 @@ fu_usb_device_open(FuDevice *device, GError **error)
+	for (gsize off = 0; off < bufsz; off += sizeof(FuUsbDeviceBosPlatformCapabilityDsinfo)) {
+		FuUsbDeviceBosPlatformCapabilityDsinfo *dsinfo =
+		    g_new0(FuUsbDeviceBosPlatformCapabilityDsinfo, 1);
+		g_ptr_array_add(dsinfos, dsinfo);
+		if (!fu_memread_uint32_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, platform_ver),
+			&dsinfo->platform_ver,
+			G_LITTLE_ENDIAN,
+			error))
+			return FALSE;
+		if (!fu_memread_uint16_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, total_length),
+			&dsinfo->total_length,
+			G_LITTLE_ENDIAN,
+			error))
+			return FALSE;
+		if (!fu_memread_uint8_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, vendor_code),
+			&dsinfo->vendor_code,
+			error))
+			return FALSE;
+		if (!fu_memread_uint8_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, alt_code),
+			&dsinfo->alt_code,
+			error))
+			return FALSE;
+		g_debug("PlatformVersion=0x%08x, TotalLength=0x%04x, VendorCode=0x%02x, "
+			"AltCode=0x%02x",
+			dsinfo->platform_ver,
+			dsinfo->total_length,
+			dsinfo->vendor_code,
+			dsinfo->alt_code);
+	}
+
+	/* sort by platform_ver, highest first */
+	g_ptr_array_sort(dsinfos, fu_usb_device_sort_by_dsinfo_platform_version_cb);
+
+	/* use Windows 8.1 as a baseline */
+	if (g_strcmp0(guid, USB_DEVICE_PLATFORM_CAPABILITY_UUID_MICROSOFT) == 0) {
+		for (guint i = 0; i < dsinfos->len; i++) {
+			FuUsbDeviceBosPlatformCapabilityDsinfo *dsinfo =
+			    g_ptr_array_index(dsinfos, i);
+			if (dsinfo->platform_ver >= 0x06030000) {
+				g_autoptr(GBytes) ds20 =
+				    fu_usb_device_fetch_platform_capability_ds20(self,
+										 dsinfo,
+										 error);
+				if (ds20 == NULL)
+					return FALSE;
+				if (!fu_usb_device_parse_platform_capability_ds20_ms(self,
+										     ds20,
+										     error)) {
+					g_prefix_error(error, "failed to parse Microsoft ds20: ");
+					return FALSE;
+				}
+				return TRUE;
+			}
+		}
+	}
+
+	/* fwupd is just quirk data */
+	if (g_strcmp0(guid, USB_DEVICE_PLATFORM_CAPABILITY_UUID_FWUPD) == 0) {
+		guint32 fwupd_version = (((guint32)FU_MAJOR_VERSION) << 16) |
+					(((guint32)FU_MINOR_VERSION) << 8) | FU_MICRO_VERSION;
+		for (guint i = 0; i < dsinfos->len; i++) {
+			FuUsbDeviceBosPlatformCapabilityDsinfo *dsinfo =
+			    g_ptr_array_index(dsinfos, i);
+
+			/* not valid */
+			if (dsinfo->platform_ver <
+			    USB_DEVICE_PLATFORM_CAPABILITY_VERSION_FWUPD_LOWEST) {
+				g_set_error(
+				    error,
+				    G_IO_ERROR,
+				    G_IO_ERROR_NOT_SUPPORTED,
+				    "invalid platform version 0x%08x, expected >= 0x%08x",
+				    dsinfo->platform_ver,
+				    (guint)USB_DEVICE_PLATFORM_CAPABILITY_VERSION_FWUPD_LOWEST);
+				return FALSE;
+			}
+
+			/* dwVersion is effectively the minimum version of fwupd */
+			if (dsinfo->platform_ver <= fwupd_version) {
```

Comment:
rather than run through all of them and parse each one - why not find the one with the biggest matching version and just parse once?

Reply:
Or at least only fail on errors if they ALL fail to parse.  I think it's painting into a corner if you try to parse even older versions and say we have a quirk schema change in newer versions.

Reply:
> why not find the one with the biggest matching version and just parse once?

Isn't that what it's doing now with the g_ptr_array_sort and the `return TRUE` in the middle of the loop?

Reply:
Oh! I missed that up above, yeah carry on.